zcash_client_sqlite/src/wallet/init.rs

```diff
@@ -324,16 +324,219 @@ pub fn init_wallet_db<
+///
+/// # Examples
+///
+/// ```
+/// # use std::error::Error;
+/// # use secrecy::SecretVec;
+/// # use tempfile::NamedTempFile;
+/// use zcash_protocol::consensus::Network;
+/// use zcash_client_sqlite::{
+///     WalletDb,
+///     util::SystemClock,
+///     wallet::init::{WalletMigrationError, WalletMigrator},
+/// };
+///
+/// # fn main() -> Result<(), Box<dyn Error>> {
+/// # let data_file = NamedTempFile::new().unwrap();
+/// # let get_data_db_path = || data_file.path();
+/// # let load_seed = || -> Result<_, String> { Ok(SecretVec::new(vec![])) };
+/// let mut db = WalletDb::for_path(get_data_db_path(), Network::TestNetwork, SystemClock)?;
+/// match WalletMigrator::new().init_or_migrate(&mut db) {
+///     Err(e)
+///         if matches!(
+///             e.source().and_then(|e| e.downcast_ref()),
+///             Some(&WalletMigrationError::SeedRequired)
+///         ) =>
+///     {
+///         let seed = load_seed()?;
+///         WalletMigrator::new()
+///             .with_seed(seed)
+///             .init_or_migrate(&mut db)
+///     }
+///     res => res,
+/// }?;
+/// # Ok(())
+/// # }
+/// ```
+pub struct WalletMigrator {
+    seed: Option<SecretVec<u8>>,
+    verify_seed_relevance: bool,
+    external_migrations: Option<Vec<Box<dyn RusqliteMigration<Error = WalletMigrationError>>>>,
+}
+
+impl Default for WalletMigrator {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl WalletMigrator {
+    /// Constructs a new wallet migrator.
+    pub fn new() -> Self {
+        Self {
+            seed: None,
+            verify_seed_relevance: true,
+            external_migrations: None,
+        }
+    }
+
+    /// Sets the seed for the migrator to use.
+    pub fn with_seed(mut self, seed: SecretVec<u8>) -> Self {
+        self.seed = Some(seed);
+        self
+    }
+
+    /// API for internal test usage only.
+    #[cfg(test)]
+    pub(crate) fn ignore_seed_relevance(mut self) -> Self {
+        self.verify_seed_relevance = false;
+        self
+    }
+
+    /// Sets the external migration graph to apply alongside the internal migrations.
+    ///
+    /// From a data management perspective, it can be useful to store additional data
+    /// alongside the `zcash_client_sqlite` wallet database. This method enables you to
+    /// provide an external [`schemerz`] migration graph that the migrator will apply to
+    /// the wallet database.
+    ///
+    /// # WARNING
+    ///
+    /// **DO NOT** depend on internal details of the `zcash_client_sqlite` schema!
+    ///
+    /// The internal migrations are written to take into account internal relationships
+    /// between the `zcash_client_sqlite` tables, but they will never take into account
+    /// external tables. In particular, this means that you **MUST NOT**:
+    /// - Modify the structure or contents of any internal table.
+    /// - Create foreign keys in your tables pointing to internal tables.
```

Comment:
This is excessively restrictive; while I understand your desire to protect the external migrations from error, that's not our job here. We need to simply document that primary keys should not be expected to be stable, and external schema should take this into account; this means that if an external table references an internal table via a foreign key, if that key changes it may require a corresponding external migration to fix. This isn't a problem, though, since all migrations run with foreign key enforcement turned off, and an external migration can fix the issue.

Reply:
We agreed that I will remove this bullet point.